nemo/collections/tts/data/vocoder_dataset.py

```diff
@@ -175,28 +238,294 @@ def __getitem__(self, index):
+            self.feature_processors = list(feature_processors.values())
+        else:
+            self.feature_processors = []
+
+        self.data_samples = []
+        self.sample_weights = []
+        self.audio_tar_filepaths = []
+        for dataset_name, dataset_info in dataset_meta.items():
+            audio_tar_filepaths = dataset_info['audio_tar_filepaths']
+            self.audio_tar_filepaths += audio_tar_filepaths
+            dataset = DatasetMeta(**dataset_info)
+            samples, weights = VocoderDataset._preprocess_manifest(
+                dataset_name=dataset_name, dataset=dataset, min_duration=min_duration, max_duration=max_duration,
+            )
+            self.data_samples += samples
+            self.sample_weights += weights
+        self.mapping = {}
+        for idx, sample in enumerate(self.data_samples):
+            file_id = os.path.splitext(os.path.basename(sample.manifest_entry["audio_filepath"]))[0]
+            if file_id not in self.mapping:
+                self.mapping[file_id] = sample
+            else:
+                raise ValueError(
+                    f"Duplicate file_id {file_id} found in manifest {sample.manifest_entry['audio_filepath']}"
+                )
+
+        logging.info(f"world size: {world_size}")
+        audio_tar_filepaths = expand_sharded_filepaths(
+            sharded_filepaths=audio_tar_filepaths,
+            global_rank=global_rank,
+            world_size=world_size,
+            shard_strategy=shard_strategy,
+        )
+
+        self._dataset = wd.WebDataset(audio_tar_filepaths, nodesplitter=None)
+
+        if shuffle_n > 0:
+            self._dataset = self._dataset.shuffle(shuffle_n)
+        else:
+            logging.info("WebDataset will not shuffle data. Consider setting shuffle_n > 0.")
+
+        self._dataset = (
+            self._dataset.rename(audio=VALID_FILE_FORMATS, key='__key__')
+            .to_tuple('audio', 'key')
+            .pipe(self._filter)
+            .map(f=self._build_sample)
+        )
+
+    def _filter(self, iterator):
+        class FilteredIterator:
+            def __init__(self, mapping):
+                self.iterator = iterator
+                self.mapping = mapping
+
+            def __iter__(self):
+                return self
+
+            def __next__(self):
+                while True:
+                    audio_bytes, audio_filename = next(self.iterator)
+                    file_id = os.path.splitext(os.path.basename(audio_filename))[0]
+                    if file_id in self.mapping:
+                        return audio_bytes, audio_filename
+
+        return FilteredIterator(self.mapping)
+
+    def _build_sample(self, tup):
+        audio_bytes, audio_filename = tup
+        file_id = os.path.splitext(os.path.basename(audio_filename))[0]
+        data = self.mapping[file_id]
+
+        audio_array, _ = sf.read(file=io.BytesIO(audio_bytes), dtype='float32')
```

Comment:
It looks like webdataset requires reading entire audio file off disk even if we only want a very small sample from it. Is there a risk of an I/O bottleneck if we have to read data on the order of 20 seconds long just to sample <= 1 second of it?

Reply:
Very good point. As far as I know, we cannot change the core indexing of WebDataset, which reads the tar file and returns the bytes.

Reply:
I don't think we can access the original file when using Webdataset. You can only access the content as bytes, at which point I/O to read the full file has already happened.

If this ends up problematic, the best solution may be to segment large files into smaller files during in the tar script.